text/0000-named-arguments.md

```diff
@@ -0,0 +1,552 @@
+http.get("https://www.rust-lang.org/", None);
+
+Window::new("Hello", 20, 50, 500, 250);
+
+free_fall(100.0, 0.0, 9.81);
+```
+
+Those are all functions that I could see written in Rust. Someone calling those functions has to remember the order in which the arguments appear. And at first sight, they can't really tell what the arguments stand for.
+
+A common workaround is to use longer function names, e.g. `get_with_timeout`, `new_with_bounds`, `free_fall_with_z0_and_v0_and_g`. However, function names that long are often considered ugly, and in the case of `new_with_bounds`, the order of arguments is still unclear.
+
+Now let’s consider calling the same functions with named arguments:
+
+```rust
+http.get("https://www.rust-lang.org/", .timeout = None);
+
+Window::new(.title = "Hello", .x = 20, .y = 50, .width = 500, .height = 250);
+
+free_fall(.z0 = 100.0, .v0 = 0.0, .g = 9.81);
+```
+
+It is now very clear what the arguments stand for, which makes maintaining such code easier.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Each function argument can be _named_ or _positional_. Positional arguments are identified by their argument position in function calls, named arguments are identified by their name:
+
+```rust
+positional_args(1, 42, "world", true);
+
+named_args(.x = 1, .y = 42, .hello = "world", .b = true);
+```
+
+## Function definition
+
+A named argument has an _argument name_ and an _argument pattern_. The argument name is part of the function's API, it's what is used _outside_ of the function. In contrast, the argument pattern is used _within_ the function.
+
+Usually, the argument pattern is the same as the argument name. The following function has a named argument `.x`, which is bound to a variable `x`:
+
+```rust
+fn identity(.x: i32) -> i32 { x }
+```
+
+It is also possible to specify a different pattern. This pattern can be used to rename or destructure the argument, for example:
+
+```rust
+fn add(.scalar a: i32, .to (b, c): (i32, i32)) -> i32 {
+    a + b + c
+}
+
+add(.scalar = 7, .to = (14, 21));
+```
+
+If a function has both positional and named arguments, the named arguments must come _after_ the positional arguments:
+
+```rust
+fn good(a: i32, .b: i32, .c: i32);
+
+fn bad(.a: i32, b: i32, .c: i32); // error!
+```
+
+The `self` argument of a method can't be a named argument.
+
+If the function is in a trait implementation, it must have the same argument names as in the trait, but the argument patterns can differ:
+
+```rust
+trait Trait {
+    fn f(.arg: i32);
+}
+impl Trait for () {
+    fn f(.arg _: i32) {}
+}
+```
+
+## Function call
+
+When calling a function, named arguments can be specified with their name. However, named arguments can't be followed by positional arguments. For example:
+
+```rust
+fn foo(a: i32, .b: i32, .c: i32) {}
+
+foo(.a = 1, .b = 2, .c = 3); // ERROR! 1st argument is not named
+foo(     1, .b = 2, .c = 3); // ok
+foo(     1,      2, .c = 3); // ok
```

Comment:
Can this flexibility *at the call site* be motivated? The motivation for names being optional at the call site is so that names can be added to existing APIs. This would be served by an all or nothing requirement at the call site, however. Is the idea that an API might initially not define a name for the first argument, but later decide that it wants one? Or that APIs might define a name for the first argument where one isn't needed, so the call sites should have the option to leave it off? I'm not sure allowing these cases warrants having quite so many different ways to call the same function.

Reply:
Otherwise we have a situation, where making a positional argument named in a function that has no named arguments is not a breaking change, because nobody could be calling it with named arguments yet; but making an argument named in a function that already has some names is, because you might have users calling using those names, and those calls will become partial calls. It is true that the first case is important so that introducing this feature doesn't break APIs, and the second one not; but still this seems like a strange inconsistency.

IMHO, removing call-site flexibility is only tenable if the function definition is required to have all-or-none (in which case, once you add names, you can't add any more names).

Reply:
Actually, the reason was that I wanted to allow specifying only some of the argument names. However, I think that positional args _after_ named arguments should be forbidden, to be forwards compatible with optional args and named args in arbitrary order.

Currently, this means that named arguments can be added backwards compatible from right to left, although that is not the primary motivation:

```rust
fn foo(a: i32, b: i32, c: i32);
fn foo(a: i32, b: i32, .c: i32);
fn foo(a: i32, .b: i32, .c: i32);
fn foo(.a: i32, .b: i32, .c: i32);
```

If optional args are added in a future RFC, this will be much more flexible. So when omitting an optional argument, arguments _after_ the omitted one must be named, but arguments _before_ the omitted one can be positional.

Reply:
When I commented that we need take into consideration the concerns over optional arguments and arbitrary order you said that was a straw man.

Then what's this?

Reply:
I want to make sure that optional arguments will be compatible with this RFC, _if_ they are added in the future. However, there is no guarantee that they will be added even if this RFC is accepted. That's why I said that we should concentrate on the features proposed in this RFC.

Reply:
I think the strawman was referring to this framing:
> Suppose named arguments are allowed, soon people will ask for
Which is actually [Slippery Slope](https://en.wikipedia.org/wiki/Slippery_slope) (but still not a relevant argument). Forward compatibility with those features should be guaranteed, but the merit of this RFC should be judged on its own.

Reply:
Linking the intended behavior by the lint should be fine here.